source/validate.cpp

```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+}
+
+// TODO(umar): Move this class to another file
+class ValidationState_t {
+ public:
+  ValidationState_t(spv_diagnostic* diag)
+      : diagnostic_(diag), instruction_counter_(0) {}
+
+  spv_result_t definedIds(uint32_t id) {
+    if (defined_ids_.find(id) == std::end(defined_ids_)) {
+      defined_ids_.insert(id);
+    } else {
+      return diag(SPV_ERROR_INVALID_ID)
+             << "ID cannot be assigned multiple times";
+    }
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t forwardDeclareId(uint32_t id) {
+    unresolved_forward_ids_.insert(id);
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t removeIfForwardDeclared(uint32_t id) {
+    unresolved_forward_ids_.erase(id);
+    return SPV_SUCCESS;
+  }
+
+  size_t unresolvedForwardIdCount() const {
+    return unresolved_forward_ids_.size();
+  }
+
+  //
+  bool isDefinedId(uint32_t id) const {
+    return defined_ids_.find(id) != std::end(defined_ids_);
+  }
+
+  // Increments the instruction count. Used for diagnostic
+  int incrementInstructionCount() { return instruction_counter_++; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) const {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(instruction_counter_)},
+                                      diagnostic_, error_code);
+  }
+
+ private:
+  spv_diagnostic* diagnostic_;
+  // Tracks the number of instructions evaluated by the validator
+  int instruction_counter_;
+
+  // All IDs which have been defined
+  unordered_set<uint32_t> defined_ids_;
+
+  // IDs which have been forward declared but have not been defined
+  unordered_set<uint32_t> unresolved_forward_ids_;
+};
+
+spv_result_t SSAPass(ValidationState_t& _,
+                     function<bool(int)> can_have_forward_declared_ids,
+                     const spv_parsed_instruction_t* inst) {
+  for (int i = 0; i < inst->num_operands; i++) {
+    const spv_parsed_operand_t& operand = inst->operands[i];
+    const spv_operand_type_t& type = operand.type;
+    const uint32_t* operand_ptr = inst->words + operand.offset;
+
+    auto ret = SPV_ERROR_INTERNAL;
+    switch (type) {
+      case SPV_OPERAND_TYPE_RESULT_ID: {
+        _.removeIfForwardDeclared(*operand_ptr);
+        ret = _.definedIds(*operand_ptr);
+        break;
+      }
+      case SPV_OPERAND_TYPE_ID:
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID:
+      case SPV_OPERAND_TYPE_SCOPE_ID: {
+        if (_.isDefinedId(*operand_ptr)) {
+          ret = SPV_SUCCESS;
+        } else if (can_have_forward_declared_ids(i)) {
+          ret = _.forwardDeclareId(*operand_ptr);
+        } else {
+          ret = _.diag(SPV_ERROR_INVALID_ID) << "ID " << *operand_ptr
+                                             << " has not been defined";
+        }
+        break;
+      }
+      default:
+        ret = SPV_SUCCESS;
+        break;
+    }
+    if (SPV_SUCCESS != ret) {
+      return ret;
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+template <SpvOp OP>
```

Comment:
Needs description.  This one and the other similar ones.
The other ones will need to reference the grammar for the operand numbers.